proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+---
+author: Thomas Bagrel
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/0>).
+**After creating the pull request, edit this file again, update the number in
+the link, and delete this bold sentence.**
+
+# Primitives for zero-copy compact regions
+
+This proposal introduces new primops in GHC to enable direct data structure building in compact regions (without temporary storage in the regular garbage-collected heap), which is useful for performance as it can greatly reduce GC cost.
+
+## Motivation
+
+[Compact regions](https://github.com/ezyang/compact) have been supported in GHC since 2017; they provide a way to allocate memory in a region that can be garbage-collected all at once instead of being traversed fully by the GC at each collection cycle. Compact regions are useful in some performance-critical use-cases, especially when dealing with large long-lived datasets.
+
+So far however, there is no way to operate in a compact region directly. One has to first build data structures in the regular GC heap, and then copy them to a compact region.
+
+I propose a new compact region RTS primitive, [`compactAddHollow#`](https://gitlab.haskell.org/tweag/ghc/-/compare/master...tbagrel1%2Fdps-compact-regions-prims?from_project_id=1#451725cc4e5d443a3b7c2adcdf224840f953b7e2_3537_3537), that allocates data constructor heap objects directly into a compact region, without setting values for their fields yet. `compactAddHollow#` lets us build structures in an _outside-in_ fashion into a region, bypassing entirely the need for a copy from the GC heap.
+
+The main intent of `compactAddHollow#` is to be encapsulated in a pure, memory-safe API, where normal functions can receive _write references_, aka _destinations_, pointing to yet-uninitialized fields of the data structures in the compact region and use them to complete the structures. This can be achieved with linear types as described in [this article](https://inria.hal.science/hal-04406360/document).
+
+A real-world use case would be a program that parses a large serialized document and keep it in memory during the whole execution. Instead of being parsed into the GC heap, and maybe copied to a compact region afterwards, with this proposal, the document can be parsed directly into the compact region, reducing greatly the time spent in garbage collection. I implemented a parser of S-expression in both naive style and DPS-style-for-compact-regions that are [presented in the **Examples** section](#library-side-more-performant-s-expression-parser) The implementation that deserializes directly into the compact region is up to twice as fast as the traditional version thanks to garbage collection time saved.
```

Comment:
This is the first occurrence of “DPS” it needs a definition if you actually need it.